documentation/sphinx/source/developer-guide.rst

```diff
@@ -823,3 +823,133 @@ Loading data is a common task in any database. Loading data in FoundationDB will
+
+Conflict Detection
+~~~~~~~~~~~~~~~~~~
+
+This section will only explain conceptually how transactions are resolved in FoundationDB. The implementation will use multiple servers running the *Resolver* role and the keyspace will be sharded across them. It will also only allow to resolve transactions where the commit version will be larger than ``commitVersion - 5Million`` (around 5 seconds...).
+
+A resolver will keep a map in memory which stores the written keys of each commit version. A simpified resolver state could look like this:
+
+=======    =======
+Version    Keys
+=======    =======
+1000       a, b
+1200       f, q, c
+1210       a
+1340       t, u, x
+=======    =======
+
+Now let's assume we have a transaction with read version *1200* and the assigned commit version will be something larger than 1340 - let's say it is *1450*. In that transaction we read keys ``b, m, s`` and we want to write to ``a``. Note that we didn't read ``a`` - so we will issue a blind write. The resolver will check whether any of the read keys (``b, m, or s``) appers in any line between version *1200* and most recent version. The last write to ``b`` was at version 1000 which was before the read version. This means that transaction read the most recent value. We don't know about any recent writes to the other keys. Therefore the resolver will decide that this transaction does *NOT* conflict and it can be committed. It will then add this new write set to its internal state so that it can resolve future transactions. The new state will look like this:
+
+=======    =======
+Version    Keys
+=======    =======
+1000       a, b
+1200       f, q, c
+1210       a
+1340       t, u, x
+1450       b, s, m
+=======    =======
+
+Note that the resolver didn't use the write set at all in order to make a decision whether the transaction can commit or not. This means that blind writes (writes to keys without reading them first) will never cause a conflict. But since the resolver will then remember these writes, blind writes can cause future transactions to conflict.
+
+Error Handling
+--------------
+
+When using FoundationDB we strongly recommend users to use the retry-loop. In Python the retry loop would look this this:
+
+.. code-block:: python
+
+   tr = tr.create_transaction()
+   while True:
+       try:
+           # execute reads and writes on FDB using the tr object
+           tr.commit().wait()
+           break
+       except FDBError as e:
+           tr.on_error(e.code).wait()
+
+This is also what the transaction decoration in python does, if you pass a ``Database`` object to a decorated function. There are some interesting properies of this retry loop:
+
+* We never create a new transaction within that loop. Instead ``tr.on_error`` will create a soft reset on the transaction.
+* ``tr.on_error`` returns a future. This is because ``on_error`` will do back off to make sure we don't overwhelm the cluster.
```

Comment:
Nothing remote, it's a future because it potentially inserts a delay to enforce a backoff.

The contract is that it will rethrow an error that isn't retryable, and otherwise it will mostly reset the transaction back to its original state and return after the backoff. State that it keeps is related to the number of retries, how long the transaction has been running, a few options that persist between retries, and the current backoff duration.

I'm personally don't really like transaction.reset and instead prefer creating new transactions, so I'm not sure I would personally recommend its use here. In fact, I think Java actually doesn't allow you to reset transactions at all anymore.

Reply:
@gabikliot